soaktest/soakclient/soakclient.go

```diff
@@ -0,0 +1,343 @@
+package main
+
+/**
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"os"
+	"os/signal"
+	"soaktest"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const END_TO_END = "end.to.end."
+
+func main() {
+
+	broker := flag.String("broker", "127.0.0.1:9092", "bootstrap servers")
+	inputTopic := flag.String("inputTopic", "inputTopic",
+		"producer will deliver messages to this topic")
+	outTopic := flag.String("outTopic", "outTopic",
+		"consumer will consume messages from this topic")
+	groupId := flag.String("groupId", "groupId",
+		"the group consumer will join")
+	numPartitions := flag.Int("numPartitions", 1, "topic partition")
+	replicationFactor := flag.Int("replicationFactor", 1, "topic replication")
+	ccloudAPIKey := flag.String("ccloudAPIKey", "", "sasl username")
+	ccloudAPISecret := flag.String("ccloudAPISecret", "", "sasl password")
+
+	flag.Parse()
+
+	var wg sync.WaitGroup
+	doneChan := make(chan bool)
+	errorChan := make(chan error)
+
+	wg.Add(1)
+	go soaktest.GetRusageMetrics(END_TO_END, &wg, doneChan, errorChan)
+
+	maxDuration, err := time.ParseDuration("30s")
+	if err != nil {
+		fmt.Printf("Failed to create maxDuration with err %s\n", err)
+		os.Exit(1)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), maxDuration)
+	defer cancel()
+
+	newTopics := []kafka.TopicSpecification{
+		{Topic: *inputTopic,
+			NumPartitions:     *numPartitions,
+			ReplicationFactor: *replicationFactor},
+		{Topic: *outTopic,
+			NumPartitions:     *numPartitions,
+			ReplicationFactor: *replicationFactor}}
+
+	err = CreateTopic(ctx, newTopics, *broker, *ccloudAPIKey, *ccloudAPISecret)
+	if err != nil {
+		fmt.Printf("Failed to create topics: %s\n", err)
+		close(doneChan)
+		os.Exit(1)
+	}
+
+	wg.Add(1)
+	go producer(doneChan, *inputTopic, *broker, *ccloudAPIKey,
+		*ccloudAPISecret, &wg, doneChan, errorChan)
+
+	wg.Add(1)
+	go consumer(*outTopic, *broker, *groupId, *ccloudAPIKey,
+		*ccloudAPISecret, &wg, doneChan, errorChan)
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	select {
+	case sig := <-sigChan:
+		fmt.Printf("Signal caught: %v\n", sig)
+	case err := <-errorChan:
+		fmt.Printf("Error caught: %s\n", err)
+	}
+
```

Comment:
I think there's a deadlock here.
errorChan and sigChan are unbuffered, thus allowing only one event, subsequent events will block on the write side.
If a signal is received the above select will read that and then come down here.
If shortly after both producer and consumer writes an error event to the errorChan, one of them will block on the write since errorChan can only hold one event and nothing is reading from it.
Now we call wg.Wait() here, but since one of the consumer/producer is blocking on a channel write they will never complete their waitgroup, and the process hangs in a deadlock here.